[quant] Add quantize_per_channel in quantized_decomposed op library

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #89272
* #89271
* #89270
* #89269
* __->__ #89268
* #89248

Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_quantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc @jianyuh @raghuramank100 @jamesr66a @vkuzo @jgong5 @Xia-Weiwen @leslie-fang-intel

jerryzh168 committed:
[quant] Add quantize_per_channel in quantized_decomposed op library

Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_quantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_quantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_quantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_quantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/_decomposed.py:
```diff
@@ -94,3 +94,31 @@ def choose_qparams_tensor(input, quant_min, quant_max, dtype):
+
+    for i in range(input.size(0)):
+        res[i] = torch.clamp(
+            torch.round(input[i] * (1.0 / scales[i]) + zero_points[i]),
```
I would have expected the cast to integer to happen here so the clamp is in the integer domain, is there a reason the cast is in L124?

jerryzh168 replied:
you mean 
```
(torch.round(input[i] * (1.0 / scales[i]) + zero_points[i]).to(dtype)
```?

I think dtype conversion need to happen after clamp is done: https://github.com/pytorch/FBGEMM/blob/main/include/fbgemm/QuantUtils.h#LL86, clamp here is actually happening in the floating point domain as well.

haven't compared the effect of doing this before clamping but my guess is that we want to make sure the value is within the range of dtype before we do this dtype conversion so that no out of bounds errors would happen

vkuzo replied:
makes sense, I agree with you on the cast.

It seems like the rounding should be changed to have add the zero_point after the round?  See https://github.com/pytorch/pytorch/pull/37188 .

Since it's super easy to miss these details even with test coverage (because we often aren't going to cover all of the corner cases), how about making sure this is unified in c++ land and calling the same function from all callsites?

jerryzh168 replied:
oh right, I just checked the code, we are using non legacy: https://github.com/pytorch/pytorch/blob/master/aten/src/ATen/native/quantized/AffineQuantizerBase.cpp#L51, which is doing rounding before adding zero_point: https://github.com/pytorch/FBGEMM/blob/main/include/fbgemm/QuantUtils.h#L78-L82
I'll change the order to align with this.

for C++, I opened this issue to track it: https://github.com/pytorch/pytorch/issues/89303, can we fix it separately? this feels like larger scope since we also have fake quant as well

vkuzo replied:
Sure, sounds reasonable to fix this later since this path is separated from the current production path.  It would be good to make sure this gets tracked and fixed before any production models are using the new path.

jerryzh168 committed:
Update on "[quant] Add quantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_quantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_quantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

vkuzo commented on torch/ao/quantization/fx/_decomposed.py:
```diff
@@ -94,3 +94,32 @@ def choose_qparams_tensor(input, quant_min, quant_max, dtype):
+    "int quant_min, int quant_max, ScalarType dtype) -> Tensor")
+
+@impl(quantized_decomposed_lib, "quantize_per_channel", "CompositeExplicitAutograd")
+def quantize_per_channel(input, scales, zero_points, axis, quant_min, quant_max, dtype):
```
do we plan to have any documentation for functions in this library?

jerryzh168 replied:
yeah, I can add some docs

jerryzh168 replied:
added this in a separate PR to avoid the CI wait time for this PR

jerryzh168 committed:
Update on "[quant] Add quantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_quantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_quantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant] Add quantize_per_channel in quantized_decomposed op library"


Summary:
att

Test Plan:
python test/test_quantization.py -k test_decomposed_quantize_per_channel

Reviewers:

Subscribers:

Tasks:

Tags:

cc jianyuh raghuramank100 jamesr66a vkuzo jgong5 Xia-Weiwen leslie-fang-intel

[ghstack-poisoned]

jerryzh168 commented:
@pytorchbot merge

jerryzh168 commented:
@pytorchbot merge
